### PR TITLE
Functionnal testing based on examples

### PR DESCRIPTION
**Pull request by peppsac**
Here's a PR which implements some basic functionnal testing using examples.

Examples are run using `mocha` + `node`, and some basic checks are implemented for each example.

The point is to basically make sure iTowns still work on each example in a quick way, even if we don't exercise the full stack.

WDYT?

### CODE REVIEW

**Comment by peppsac** [comment]
I forgot something: you can test using the new `npm run test-examples` command :)

**Comment by peppsac** [comment]
It seems that nobody is interested in testing... I should have used a fork of an obscure test framework instead of the standard one; I'm sure it would have generated a lot of comments :-)

**Comment by gchoqueux** [comment]
yes, I'am interesting, I will be regarding soon
I have too a PR on test #275

File: examples/planar.js [hunk 1]
```diff
@@ -0,0 +1,76 @@
+// # Planar (EPSG:3946) viewer
```
Comment (gchoqueux) [inline examples/planar.js:2]: in browser `/examples/planar.html` have error 
`Uncaught ReferenceError: bbox is not defined`

File: examples/postprocessing.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    r.setViewport(0, 0, g.getWindowSize().x, g.getWindowSize().y);
+    r.render(
+        postprocessScene,
+        cam);
+};
+
+itowns.Fetcher.json('/examples/layers/JSONLayers/Ortho.json').then(result => globeView.addLayer(result));
+itowns.Fetcher.json('/examples/layers/JSONLayers/IGN_MNT.json').then(result => globeView.addLayer(result));
+
+exports.globeView = globeView;
```
Comment (gchoqueux) [inline examples/postprocessing.js:52]: error `exports` not define in browser

File: test/externalscene_test.js [hunk 3]
```diff
@@ -0,0 +1,14 @@
+var itownsTesting = require('./itowns-testing.js');
+var example = require('../examples/externalscene.js');
+
+describe('External Scene', function() {
+    it('should use the user constructed scene', function(done) {
+        example.globeView.mainLoop.addEventListener('command-queue-empty', () => {
```
Comment (gchoqueux) [inline test/externalscene_test.js]: You could use this event?
`example.globeView.addEventListener(itowns.GLOBE_VIEW_EVENTS.GLOBE_INITIALIZED`...

File: test/globe_test.js [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+            for (var obj of example.view.wgs84TileLayer.level0Nodes) {
+                itownsTesting.countVisibleAndDisplayed(obj);
+            }
+
+            if (initialState) {
+                initialStateTest();
+                initialState = false;
+                example.view.controls.setRange(10000, false);
+                example.view.notifyChange(true);
```
Comment (gchoqueux) [inline test/globe_test.js]: why you use `notifyChange`?

File: examples/externalscene.js [hunk 5]
```diff
@@ -0,0 +1,19 @@
+const positionOnGlobe = { longitude: 2.351323, latitude: 48.856712, altitude: 25000000 };
+
+const scene = new itowns.THREE.Scene();
+
+// iTowns namespace defined here
+const viewerDiv = document.getElementById('viewerDiv');
+viewerDiv.appendChild(renderer.domElement);
```
Comment (gchoqueux) [inline examples/externalscene.js]: in browser `renderer` is undefined

File: examples/globe.js [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+// These will deform iTowns globe geometry to represent terrain elevation.
+itowns.Fetcher.json('/examples/layers/JSONLayers/IGN_MNT.json').then(result => globeView.addLayer(result));
+
+// Listen for globe full initialisation event
+globeView.addEventListener(itowns.GLOBE_VIEW_EVENTS.GLOBE_INITIALIZED, () => {
+    /* eslint-disable-next-line no-console */
+    console.info('Globe initialized');
+});
+
+exports.view = globeView;
```
Comment (gchoqueux) [inline examples/globe.js]: `exports` is undefined in browser

**Review by gchoqueux — CHANGES_REQUESTED** [review]

File: test/externalscene_test.js (hunk 3 above)
Comment (peppsac) [reply test/externalscene_test.js]: Hmm for no particular reason, both are equivalent.

**Comment by peppsac** [comment]
Fixed all your comments.

**Review by gchoqueux — APPROVED** [review]

**Comment by gchoqueux** [comment]
needs rebasing

**Comment by peppsac** [comment]
The tests have a different behavior on Travis (failure).
I'm trying to figure out what causes the difference - let me know if you have any ideas (or if the tests fail on your computer).

**Comment by tbroyer** [comment]
Same error on my machine (node v8.0.0, npm v5.0.0)

**Comment by peppsac** [comment]
Thanks I can reproduce it here with the same node version and I think I understood the root issue.

**Comment by tbroyer** [comment]
FWIW, those tests are non-deterministic. Here's the same test run 3 times in a row:
```
$ mocha test/globe_test.js


  Globe example
Globe initialized
(node:13887) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 1): TypeError: this.renderer.setScissor is not a function
(node:13887) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
(node:13887) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 2): TypeError: this.material.setTexturesLayer is not a function
(node:13887) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 3): TypeError: this.material.setTexturesLayer is not a function
(node:13887) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 4): TypeError: this.material.setTexturesLayer is not a function
(node:13887) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 5): TypeError: this.material.setTexturesLayer is not a function
    ✓ should subdivide like expected (137ms)


  1 passing (252ms)

$ mocha test/globe_test.js


  Globe example
Globe initialized
(node:13911) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 1): TypeError: this.renderer.setScissor is not a function
(node:13911) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
    ✓ should subdivide like expected (140ms)


  1 passing (255ms)

$ mocha test/globe_test.js


  Globe example
Globe initialized
(node:13932) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 1): TypeError: this.renderer.setScissor is not a function
(node:13932) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
    1) should subdivide like expected


  0 passing (2s)
  1 failing

  1) Globe example should subdivide like expected:
     Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.
```

**Comment by peppsac** [comment]
From the beginning I've made these tests more and more deterministic: I won't merge tests that sometimes work, sometimes don't.

I'll push a new version soon, let me know if they work better.

**Comment by peppsac** [comment]
Should be fixed now.
@tbroyer let me know if it's still unreliable on your setup, thanks!

**Comment by tbroyer** [comment]
Appears to work reliably now.

**Pull request merged**